### ISSUE TITLE
Breakage of stable version of kernel prevents release of LTS version.

### ISSUE THREAD

**Issue by GregoryLand** [archzfs/archzfs#560]
Right now the automatic builds are broken. https://github.com/archzfs/archzfs/actions/runs/12002432031

```
make  all-recursive
make[1]: Entering directory '/build/zfs-linux/src/zfs-2.2.6'
Making all in include
make[2]: Entering directory '/build/zfs-linux/src/zfs-2.2.6/include'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/build/zfs-linux/src/zfs-2.2.6/include'
Making all in module
make[2]: Entering directory '/build/zfs-linux/src/zfs-2.2.6/module'
mkdir -p os/linux/spl/
mkdir -p avl/ icp/ icp/algs/aes/ icp/algs/blake3/ icp/algs/edonr/ icp/algs/modes/ icp/algs/sha2/ icp/algs/skein/ icp/api/ icp/asm-aarch64/blake3/ icp/asm-aarch64/sha2/ icp/asm-arm/sha2/ icp/asm-ppc64/blake3/ icp/asm-ppc64/sha2/ icp/asm-x86_64/aes/ icp/asm-x86_64/blake3/ icp/asm-x86_64/modes/ icp/asm-x86_64/sha2/ icp/core/ icp/io/ icp/spi/ lua/ lua/setjmp/ nvpair/ os/linux/zfs/ unicode/ zcommon/ zfs/ zstd/ zstd/lib/common/ zstd/lib/compress/ zstd/lib/decompress/
make -C /usr/lib/modules/6.12.1-arch1-1/build  \
	  \
	M="$PWD"  CONFIG_ZFS=m modules
make[3]: Entering directory '/usr/lib/modules/6.12.1-arch1-1/build'
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-atomic.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-condvar.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-cred.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-err.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-generic.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-kmem-cache.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-kmem.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-kstat.o
In file included from /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-generic.c:31:
/build/zfs-linux/src/zfs-2.2.6/include/os/linux/spl/sys/kmem_cache.h:205:9: warning: "kmem_cache_create" redefined
  205 | #define kmem_cache_create(name, size, align, ctor, dtor, rclm, priv, vmp, fl) \
      |         ^~~~~~~~~~~~~~~~~
In file included from ./include/linux/fs.h:45,
                 from ./include/linux/compat.h:17,
                 from ./arch/x86/include/asm/ia32.h:7,
                 from ./arch/x86/include/asm/elf.h:10,
                 from ./include/linux/elf.h:6,
                 from ./include/linux/module.h:19,
                 from /build/zfs-linux/src/zfs-2.2.6/include/os/linux/spl/sys/sysmacros.h:27,
                 from /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-generic.c:27:
./include/linux/slab.h:404:9: note: this is the location of the previous definition
  404 | #define kmem_cache_create(__name, __object_size, __args, ...)           \
      |         ^~~~~~~~~~~~~~~~~
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-proc.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-procfs-list.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-shrinker.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-taskq.o
In file included from /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-proc.c:32:
/build/zfs-linux/src/zfs-2.2.6/include/os/linux/spl/sys/kmem_cache.h:205:9: warning: "kmem_cache_create" redefined
  205 | #define kmem_cache_create(name, size, align, ctor, dtor, rclm, priv, vmp, fl) \
      |         ^~~~~~~~~~~~~~~~~
In file included from ./include/linux/fs.h:45,
                 from ./include/linux/compat.h:17,
                 from ./arch/x86/include/asm/ia32.h:7,
                 from ./arch/x86/include/asm/elf.h:10,
                 from ./include/linux/elf.h:6,
                 from ./include/linux/module.h:19,
                 from /build/zfs-linux/src/zfs-2.2.6/include/os/linux/spl/sys/kstat.h:27,
                 from /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-proc.c:30:
./include/linux/slab.h:404:9: note: this is the location of the previous definition
  404 | #define kmem_cache_create(__name, __object_size, __args, ...)           \
      |         ^~~~~~~~~~~~~~~~~
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-trace.o
  CC [M]  /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-tsd.o
/build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.c: In function ‘issig’:
/build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.c:186:28: error: passing argument 1 of ‘dequeue_signal’ from incompatible pointer type [-Wincompatible-pointer-types]
  186 |         if (dequeue_signal(task, &set, &__info) != 0) {
      |                            ^~~~
      |                            |
      |                            struct task_struct *
In file included from ./include/linux/rcuwait.h:6,
                 from ./include/linux/percpu-rwsem.h:7,
                 from ./include/linux/fs.h:33,
                 from ./include/linux/compat.h:17,
                 from ./arch/x86/include/asm/ia32.h:7,
                 from ./arch/x86/include/asm/elf.h:10,
                 from ./include/linux/elf.h:6,
                 from ./include/linux/module.h:19,
                 from /build/zfs-linux/src/zfs-2.2.6/include/os/linux/spl/sys/thread.h:27,
                 from /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.c:26:
./include/linux/sched/signal.h:279:37: note: expected ‘sigset_t *’ but argument is of type ‘struct task_struct *’
  279 | extern int dequeue_signal(sigset_t *mask, kernel_siginfo_t *info, enum pid_type *type);
      |                           ~~~~~~~~~~^~~~
/build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.c:186:34: error: passing argument 2 of ‘dequeue_signal’ from incompatible pointer type [-Wincompatible-pointer-types]
  186 |         if (dequeue_signal(task, &set, &__info) != 0) {
      |                                  ^~~~
      |                                  |
      |                                  sigset_t *
./include/linux/sched/signal.h:279:61: note: expected ‘kernel_siginfo_t *’ {aka ‘struct kernel_siginfo *’} but argument is of type ‘sigset_t *’
  279 | extern int dequeue_signal(sigset_t *mask, kernel_siginfo_t *info, enum pid_type *type);
      |                                           ~~~~~~~~~~~~~~~~~~^~~~
/build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.c:186:40: error: passing argument 3 of ‘dequeue_signal’ from incompatible pointer type [-Wincompatible-pointer-types]
  186 |         if (dequeue_signal(task, &set, &__info) != 0) {
      |                                        ^~~~~~~
      |                                        |
      |                                        spl_kernel_siginfo_t * {aka struct kernel_siginfo *}
./include/linux/sched/signal.h:279:82: note: expected ‘enum pid_type *’ but argument is of type ‘spl_kernel_siginfo_t *’ {aka ‘struct kernel_siginfo *’}
  279 | extern int dequeue_signal(sigset_t *mask, kernel_siginfo_t *info, enum pid_type *type);
      |                                                                   ~~~~~~~~~~~~~~~^~~~
make[5]: *** [scripts/Makefile.build:229: /build/zfs-linux/src/zfs-2.2.6/module/os/linux/spl/spl-thread.o] Error 1
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [/usr/lib/modules/6.12.1-arch1-1/build/Makefile:1936: /build/zfs-linux/src/zfs-2.2.6/module] Error 2
make[3]: *** [Makefile:224: __sub-make] Error 2
make[3]: Leaving directory '/usr/lib/modules/6.12.1-arch1-1/build'
make[2]: *** [Makefile:56: modules-Linux] Error 2
make[2]: Leaving directory '/build/zfs-linux/src/zfs-2.2.6/module'
make[1]: *** [Makefile:12160: all-recursive] Error 1
make[1]: Leaving directory '/build/zfs-linux/src/zfs-2.2.6'
```

zfs-2.2.6 will not build against the current stable kernel.  This breaks the build system and keeps it from attempting to build against the LTS version.

**Comment by n0xena** [comment]
I got a build done: https://github.com/n0xena/archzfs/releases/tag/2.2.6_2.3.0-rc3_6.11.9
how: just build utils and lts manual without std

on a clean VM with clean-chroot-manager installed
```
sudo ccm64 n
sudo ccm64 c
<archlinux>$ sudo ./build.sh utils update
<archlinux>$ sudo ./build.sh utils make
<archlinux>$ sudo ./build.sh lts update
<archlinux>$ sudo ./build.sh lts make
```
make sure to set your correct username in conf.sh before using any build.sh command

**Comment by GregoryLand** [comment]
Yea.  Right now the github build automation runs each kernel in sequence as part of the docker container entrypoint. 

build-container/entrypoint.sh
```build utils
build std
build lts
build hardened
build zen
build dkms```

The script entrypoint.sh has `set -e` at line 2 of the script.  So if any of the kernel builds fail it halts the build.

I see two possible solutions to this.
1) Remove `set -e` and let the rest of the script run.
2) Break utils out into its own build stage and then have std, lts, hardened, zen and dkms run as there own builds dependent on utils.

LTS not getting built because std fails to build has been a long standing issue. It has gotten worse since openzfs upstream added explicit enforcement checks for kernel versions.

**Comment by Doridian** [comment]
> 1. Remove `set -e` and let the rest of the script run. I do not know how this would play out with the github releases.  Would this result in a build that didn't have packages for std or would the release contain the packages from the previous release? @Doridian

@GregoryLand It would result in a new repo without any of the failed packages in it. If we decide to push forward with such a change, it should be a fairly small change to first download the old repo/packages such that we can preserve them, however.

**Comment by GregoryLand** [comment]
@Doridian Should I take a stab at it?  I don't want to step on your toes if its something you can do pretty quickly.

**Comment by n0xena** [comment]
To continue to already quite exhaustive discussion about what to support:
From what I got (please correct me if I'm wrong) the consent was agree upon: "the 'supported' version should match what upstream zfs and arch combine support - hence: pretty much just LTS".
So the way I see it: as LTS is the preferred target anyway - make it a priority to get at least utils+lts build - and anything else (std, zen, hardened) should be considered unstable or even experimental anyways.
So maybe swap the order to

```
build utils
build lts
...
anything else
```
could provide at least LTS to fulfil this overall "stay on LTS" scheme

but - that's just my personal opinion
for anybody who dares to try - 2.3.0-rc3 on 6.12.1 works smooth so far for me

**Comment by GregoryLand** [comment]
The build would still fail to upload anything since adding to the repos/signing occurs after all the builds have been finished.  So changing the ordering of the builds inside the script would not change the output.

**Comment by GregoryLand** [comment]
@Doridian So I took a quick stab at those changes you recommended.  Did a test run of the automation on my git repo and it worked as expected.  This was just a quick and dirty set of commits to get the repo building again while we work out the issues.  The build will fail to report failures with the quick pull request I made.

**Comment by r-maerz** [comment]
> > 1. Remove `set -e` and let the rest of the script run. I do not know how this would play out with the github releases.  Would this result in a build that didn't have packages for std or would the release contain the packages from the previous release? @Doridian
> 
> @GregoryLand It would result in a new repo without any of the failed packages in it. If we decide to push forward with such a change, it should be a fairly small change to first download the old repo/packages such that we can preserve them, however.

@Doridian Are you open to a complete rebuild of the pipeline? I'd set it up in a way that all kernel packages can be build independently and in parallel and targeting different releases (experimental vs stable) would also be supported per package. In case of build failures, previous versions of the relevant kernel package would be retained.

What I have in mind is quite different from what's currently being used in this project, though. So I am reluctant to invest the time if there's a high chance a PR would be declined later on due to that fact.

**Comment by Doridian** [comment]
> @Doridian Are you open to a complete rebuild of the pipeline? I'd set it up in a way that all kernel packages can be build independently and in parallel and targeting different releases (experimental vs stable) would also be supported per package. In case of build failures, previous versions of the relevant kernel package would be retained.
> 
> What I have in mind is quite different from what's currently being used in this project, though. So I am reluctant to invest the time if there's a high chance a PR would be declined later on due to that fact.

@r-maerz I am open to rewriting it if that helps.
One thing to keep in mind if you want to take this on is that we will only have 2 repos (stable, and experimental, maybe more later), so even if you split out the build tasks into individual steps with, say, a `matrix` build, you'd still need a step in the end to collect all artifacts, augment it with the previous artifacts from the repo on failed builds, and then regenerate a new repo.

Not sure if splitting up the task on GItHub's end vs  just running the script with `-e` disabled and then figuring out what failed is easier or better. We can always use [GitHub's checks API](https://docs.github.com/en/rest/checks) to report the status of individual packages, even in a single step build.
But it's up to you, I'm looking forward to what you can come up with :)

**Comment by n0xena** [comment]
2.2.7 with 6.12 support released https://github.com/openzfs/zfs/releases/tag/zfs-2.2.7

**Comment by Doridian** [comment]
Closing as duplicate of https://github.com/archzfs/archzfs/issues/506
